app/models/rubygem.rb

```diff
@@ -31,7 +31,7 @@ def create_gem_download
   end
 
   def self.with_versions
-    where("rubygems.id IN (SELECT rubygem_id FROM versions where versions.indexed IS true)")
+    where(indexed: true)
```

Comment:
The previous version used a pattern that is sadly common among the Rails community, namely selecting where the ID is in a set of IDS based on a subquery. This is far more efficient than doing with a pluck, which is good. However, an alternative approach that would not add a new indexed column would be something like:

`joins(:versions).where('versions.indexed IS true).select('DISTINCT...')...`

Did you try this to compare the query time of this approach vs the subquery approach?

Reply:
Interesting, can you please post 2 queries (current and with your proposed change) to compare?

Reply:
Nevermind. This is why we test!

Current analysis on a full select of rubygems that are indexed using the current method:

```
rubygems_clone_20200515=# EXPLAIN ANALYZE SELECT rubygems.* FROM "rubygems" WHERE (rubygems.id IN (SELECT rubygem_id FROM versions where versions.indexed IS true));
                                                                       QUERY PLAN                                                                        
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1000.43..62064.39 rows=47115 width=44) (actual time=0.728..352.425 rows=159504 loops=1)
   Workers Planned: 1
   Workers Launched: 1
   ->  Nested Loop Semi Join  (cost=0.43..56352.89 rows=27715 width=44) (actual time=0.063..328.884 rows=79752 loops=2)
         ->  Parallel Seq Scan on rubygems  (cost=0.00..2397.55 rows=100655 width=44) (actual time=0.009..6.379 rows=85556 loops=2)
         ->  Index Scan using index_versions_on_rubygem_id on versions  (cost=0.43..2.31 rows=23 width=4) (actual time=0.004..0.004 rows=1 loops=171113)
               Index Cond: (rubygem_id = rubygems.id)
               Filter: (indexed IS TRUE)
               Rows Removed by Filter: 0
 Planning time: 0.815 ms
 Execution time: 355.888 ms
```
The method I was thinking about using an inner join...
```
rubygems_clone_20200515=# EXPLAIN ANALYZE SELECT rubygems.* FROM rubygems INNER JOIN versions ON versions.rubygem_id = rubygems.id AND versions.indexed IS true;
                                                         QUERY PLAN                                                          
-----------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=6745.04..89495.98 rows=1062705 width=44) (actual time=34.443..475.951 rows=1061813 loops=1)
   Hash Cond: (versions.rubygem_id = rubygems.id)
   ->  Seq Scan on versions  (cost=0.00..70153.27 rows=1062705 width=4) (actual time=0.015..212.778 rows=1061813 loops=1)
         Filter: (indexed IS TRUE)
         Rows Removed by Filter: 104889
   ->  Hash  (cost=3102.13..3102.13 rows=171113 width=44) (actual time=33.616..33.616 rows=171113 loops=1)
         Buckets: 65536  Batches: 4  Memory Usage: 3459kB
         ->  Seq Scan on rubygems  (cost=0.00..3102.13 rows=171113 width=44) (actual time=0.025..12.192 rows=171113 loops=1)
 Planning time: 0.921 ms
 Execution time: 497.397 ms
```
Current method is faster. Disregard my thought.